carma_wm/src/WorldModel.cpp

```diff
@@ -0,0 +1,191 @@
+    new_points = lanelets[i].centerline2d().basicLineString();
+    centerline_points = concatenate_line_strings(centerline_points, new_points);
+  }
+
+  return centerline_points;
+}
+
+lanelet::BasicLineString2d 
+WorldModel::concatenate_line_strings(const lanelet::BasicLineString2d& a, 
+                                          const lanelet::BasicLineString2d& b)
+{
+  lanelet::BasicLineString2d out;
+  out.insert(out.end(), a.begin(), a.end());
+  out.insert(out.end(), b.begin(), b.end());
+
+  return out;
+}
+
+std::vector<Eigen::Vector2d> 
+WorldModel::compute_finite_differences(const lanelet::BasicLineString2d& data)
+{
+  std::vector<Eigen::Vector2d> out;
+  Eigen::Vector2d diff;
+  Eigen::Vector2d vec_i_plus_1;
+  Eigen::Vector2d vec_i_minus_1;
+
+  for (int i = 0; i < data.size(); i++) {
+    if (i == 0) {
+      // Compute forward derivative
+      diff = (data[i + 1] - data[i]);
+    } else if (i == data.size() - 1) {
+      // Compute backward derivative
+      diff = (data[i] - data[i - 1]);
+    } else {
+      // Else, compute centered derivative
+      vec_i_plus_1 = data[i + 1];
+      vec_i_minus_1 = data[i - 1];
+      diff = (vec_i_plus_1 - vec_i_minus_1)/2.0;
+    }
+
+    out.push_back(diff);
+  }
+
+  return out;
+}
+
+std::vector<double> 
+WorldModel::compute_finite_differences(const std::vector<double>& data)
+{
+  std::vector<double> out;
+  double diff;
+  for (int i = 0; i < data.size(); i++) {
+
+    if (i == 0) {
+      diff = data[i + 1] - data[i];
+    } else if (i == data.size() - 1) {
+      diff = data[i] - data[i - 1];
+    } else {
+      diff = (data[i + 1] - data[i - 1])/2.0;
+    }
+    
+    out.push_back(diff);
+  }
+
+  return out;
+}
+
+std::vector<Eigen::Vector2d> 
+WorldModel::compute_finite_differences(const std::vector<Eigen::Vector2d>& x,const std::vector<double>& y)
+{
+  if (x.size() != y.size()) {
+    throw std::invalid_argument("Attempting to differentiate two unequal sized lists!");
+  }
+
+  std::vector<Eigen::Vector2d> out;
+  Eigen::Vector2d diff;
+  for (int i = 0; i < x.size(); i++) {
+    if (i == 0) {
+      diff = (x[i + 1] - x[i])/(y[i + 1] - y[i]);
+    } else if (i == x.size() - 1) {
+      diff = (x[i] - x[i - 1])/(y[i] - y[i - 1]);
+    } else {
+      diff = (x[i + 1] - x[i - 1])/(y[i + 1] - y[i - 1]);
+    }
+    
+    out.push_back(diff);
+  }
+
+  return out;
+}
+
+std::vector<double>
+WorldModel::compute_arc_lengths(const lanelet::BasicLineString2d& data)
+{
+  std::vector<double> out;
+  double total = 0;
+  double diff = 0;
+  for (int i =  0; i < data.size(); i++) {
+    if (i == 0) {
+      out.push_back(0);
+    } else {
+      diff = compute_euclidean_distance(data[i - 1], data[i]);
+      total += diff;
+      out.push_back(total);
+    }
+  }
+
+  return out;
+}
+
+double 
+WorldModel::compute_euclidean_distance(const Eigen::Vector2d& a, const Eigen::Vector2d& b)
+{
+  return std::sqrt(std::pow(b[0] - a[0], 2) + std::pow(b[1] - a[1], 2));
```

Comment:
I think you want to use std::pow for b[0] and a[0] as well, right?

Reply:
I'm not sure I follow, it's already `std::pow(b[0] - a[0], 2) + std::pow(b[1] - a[1], 2)` so `(x1 - x2)^2 + (y1 - y2)^2`